src/evaluators/JSXElement.js

```diff
@@ -0,0 +1,370 @@
+}
+
+function isTagName(ast: BabelNode) {
+  return ast.type === "JSXIdentifier" && /^[a-z]|\-/.test(((ast: any): BabelNodeJSXIdentifier).name);
+}
+
+function getDefaultProps(
+  elementType: BabelNodeJSXIdentifier | BabelNodeJSXMemberExpression,
+  env,
+  realm: Realm
+): null | ObjectValue {
+  let name;
+  if (elementType.type === "JSXIdentifier") {
+    name = elementType.name;
+  }
+  if (!isTagName(elementType) && typeof name === "string") {
+    try {
+      // find the value of "ComponentXXX.defaultProps"
+      let defaultProps = GetValue(
+        realm,
+        env.evaluate(t.memberExpression(t.identifier(name), t.identifier("defaultProps")), false)
+      );
+
+      if (defaultProps instanceof ObjectValue) {
+        return defaultProps;
+      }
+    } catch (error) {
+      return null;
+    }
+  }
+  return null;
+}
+
+function evaluateJSXChildren(children: Array<BabelNode>, strictCode: boolean, env: LexicalEnvironment, realm: Realm) {
+  if (children.length === 0) {
+    return null;
+  }
+  if (children.length === 1) {
+    let singleChild = evaluateJSXValue(children[0], strictCode, env, realm);
+
+    if (singleChild instanceof StringValue) {
+      let text = cleanJSXElementLiteralChild(singleChild.value);
+      if (text !== null) {
+        singleChild.value = text;
+      }
+    }
+    return singleChild;
+  }
+  let array = ArrayCreate(realm, 0);
+  let dynamicChildrenLength = children.length;
+  let dynamicIterator = 0;
+  let lastChildValue = null;
+  for (let i = 0; i < children.length; i++) {
+    let value = evaluateJSXValue(children[i], strictCode, env, realm);
+    if (value instanceof StringValue) {
+      let text = cleanJSXElementLiteralChild(value.value);
+      if (text === null) {
+        dynamicChildrenLength--;
+        // this is a space full of whitespace, so let's proceed
+        continue;
+      } else {
+        value.value = text;
+      }
+    }
+    lastChildValue = value;
+    CreateDataPropertyOrThrow(realm, array, "" + dynamicIterator, value);
+    dynamicIterator++;
+  }
+  if (dynamicChildrenLength === 1) {
+    return lastChildValue;
+  }
+
+  Set(realm, array, "length", new NumberValue(realm, dynamicChildrenLength), false);
+  return array;
+}
+
+function evaluateJSXAttributes(
+  elementType: BabelNodeJSXIdentifier | BabelNodeJSXMemberExpression,
+  astAttributes: Array<BabelNodeJSXAttribute | BabelNodeJSXSpreadAttribute>,
+  astChildren: Array<BabelNode>,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm
+) {
+  let attributes = new Map();
+  let children = evaluateJSXChildren(astChildren, strictCode, env, realm);
+  let defaultProps = getDefaultProps(elementType, env, realm);
+
+  // defaultProps are a bit like default function arguments
+  // if an actual value exists, it should overwrite the default value
+  if (defaultProps !== null) {
+    for (let [key, defaultProp] of defaultProps.properties) {
```

Comment:
This should probably go through the normal property `Get` operations and such rather than iterating internal data structures. Since if there are getters etc. they should get invoked.